src/nvim/api/vim.c

```diff
@@ -250,32 +249,37 @@ void nvim_set_hl(Integer ns_id, String name, Dictionary val, Error *err)
 /// |nvim_set_decoration_provider| on_win and on_line callbacks
 /// are explicitly allowed to change the namespace during a redraw cycle.
 ///
-/// @param ns_id the namespace to activate
+// @param ns_id the namespace to activate
 /// @param[out] err Error details, if any
-void nvim__set_hl_ns(Integer ns_id, Error *err)
-  FUNC_API_FAST
+void nvim_set_hl_ns(Integer ns_id, Error *err)
+  FUNC_API_SINCE(7)
 {
-  if (ns_id >= 0) {
-    ns_hl_active = (NS)ns_id;
-  }
-
-  // TODO(bfredl): this is a little bit hackish.  Eventually we want a standard
-  // event path for redraws caused by "fast" events. This could tie in with
-  // better throttling of async events causing redraws, such as non-batched
-  // nvim_buf_set_extmark calls from async contexts.
-  if (!provider_active && !ns_hl_changed) {
-    multiqueue_put(main_loop.events, on_redraw_event, 0);
+  if (ns_id < 0) {
+    api_set_error(err, kErrorTypeValidation, "no such namespace");
+    return;
   }
-  ns_hl_changed = true;
+  ns_hl_global = (NS)ns_id;
+  hl_check_ns();
+  redraw_all_later(NOT_VALID);
 }
 
-static void on_redraw_event(void **argv)
-  FUNC_API_NOEXPORT
+/// Set active namespace for highlights.
+///
+/// NB: this function can be called from async contexts, but the
+/// semantics are not yet well-defined. To start with
+/// |nvim_set_decoration_provider| on_win and on_line callbacks
+/// are explicitly allowed to change the namespace during a redraw cycle.
+///
+/// @param ns_id the namespace to activate
+/// @param[out] err Error details, if any
+void nvim_set_hl_ns_fast(Integer ns_id, Error *err)
```

Comment:
don't think we should start adding "fast" suffix to api names. It's already covered by an attribute and could be part of the metadata if it isn't already.